sql/core/src/main/scala/org/apache/spark/sql/execution/Aggregate.scala

```diff
@@ -17,181 +17,461 @@
-            val currentBuffer = currentEntry.getValue
-
-            var i = 0
-            while (i < currentBuffer.length) {
-              // Evaluating an aggregate buffer returns the result.  No row is required since we
-              // already added all rows in the group using update.
-              aggregateResults(i) = currentBuffer(i).eval(EmptyRow)
-              i += 1
+        while (iter.hasNext) {
+          val currentRow = iter.next()
+
+          var idx = 0
+          while (idx < aggregateFunctionBinds.length) {
+            val ae = aggregates(idx)
+            val value = ae.eval(currentRow)
+
+            if (ae.distinct) {
+              if (value != null && !seens(idx).contains(value)) {
+                ae.iterate(value, buffer)
+                seens(idx).add(value)
+              }
+            } else {
+              ae.iterate(value, buffer)
             }
-            resultProjection(joinedRow(aggregateResults, currentGroup))
+            idx += 1
           }
         }
+
+        createIterator(aggregates, Iterator(ibs))
+      } else {
+        val results = new OpenHashMap[Row, InputBufferSeens]()
+
+        while (iter.hasNext) {
+          val currentRow = iter.next()
+
+          val keys = groupByProjection(currentRow)
+          results(keys) match {
+            case null =>
+              val buffer = new GenericMutableRow(bufferSchema.length)
+              // TODO save the memory only for those DISTINCT aggregate expressions
+              val seens = new Array[OpenHashSet[Any]](aggregateFunctionBinds.length)
+
+              var idx = 0
+              while (idx < aggregateFunctionBinds.length) {
+                val ae = aggregates(idx)
+                val value = ae.eval(currentRow)
+                ae.reset(buffer)
+                ae.iterate(value, buffer)
+
+                if (ae.distinct) {
+                  val seen = new OpenHashSet[Any]()
+                  if (value != null) {
+                    seen.add(value)
+                  }
+                  seens.update(idx, seen)
+                }
+
+                idx += 1
+              }
+              results(keys.copy()) = new InputBufferSeens(currentRow.copy(), buffer, seens)
+
+            case inputBufferSeens =>
+              var idx = 0
+              while (idx < aggregateFunctionBinds.length) {
+                val ae = aggregates(idx)
+                val value = ae.eval(currentRow)
+
+                if (ae.distinct) {
+                  if (value != null && !inputBufferSeens.seens(idx).contains(value)) {
+                    ae.iterate(value, inputBufferSeens.buffer)
+                    inputBufferSeens.seens(idx).add(value)
+                  }
+                } else {
+                  ae.iterate(value, inputBufferSeens.buffer)
+                }
+                idx += 1
+              }
+          }
+        }
+
+        createIterator(aggregates, results.iterator.map(_._2))
```

Comment:
As in the description of this PR, I am trying to improve the code by eliminating the `DistinctAggregate`, but I guess we don't want to mix the `AggregatePreShuffle` and `AggregatePostShuffle`, as the code will be very difficult to understand if we do that.